### PR TITLE
DPC++ K-Means performance fix

### PR DESCRIPTION
**Pull request by amgrigoriev**
# Description
- Removed initialization of distance matrix.
- Added computation of squares
- Modified selection to use pre-calculated squares

### CODE REVIEW

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp [hunk 1]
```diff
@@ -179,11 +187,12 @@ sycl::event kernels_fp<Float>::assign_clusters(sycl::queue& queue,
...
     ONEDAL_ASSERT(distances.get_dimension(1) >= centroids.get_dimension(0));
+    ONEDAL_ASSERT(centroid_squares.get_dimension(0) == centroids.get_dimension(0));
+    ONEDAL_ASSERT(data_squares.get_dimension(0) == data.get_dimension(0));
     sycl::event selection_event;
     const auto row_count = data.get_dimension(0);
     const auto column_count = data.get_dimension(1);
     const auto centroid_count = centroids.get_dimension(0);
-    pr::distance<Float, Metric> block_distances(queue);
     auto block_count =
         row_count / block_size_in_rows + std::int64_t(row_count % block_size_in_rows > 0);
```
Comment (makart19) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp:197]:
```suggestion
        row_count / block_size_in_rows + bool(row_count % block_size_in_rows);
```

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp [hunk 2]
```diff
@@ -93,24 +97,27 @@ std::int64_t kernels_fp<Float>::get_part_count_for_partial_centroids(sycl::queue
...
+    ONEDAL_ASSERT(centroid_squares.get_dimension(0) == distances.get_dimension(1));
 
-    const std::int64_t col_count = data.get_dimension(1);
-    const std::int64_t row_count = data.get_dimension(0);
-    const std::int64_t stride = data.get_dimension(1);
+    const std::int64_t cluster_count = distances.get_dimension(1);
+    const std::int64_t row_count = distances.get_dimension(0);
+    const std::int64_t stride = distances.get_dimension(1);
 
     const std::int64_t preffered_wg_size = 128;
```
Comment (makart19) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp:115]: Please declare it as a member of kernels_fp struct, to have all config consts in one place.
Definition it here doesn't differ from `magic number` like  bk::get_scaled_wg_size_per_row(queue, cluster_count, 128);

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp [hunk 3]
```diff
@@ -135,8 +142,8 @@ sycl::event kernels_fp<Float>::select(sycl::queue& queue,
 
                 std::int32_t index = -1;
                 Float value = fp_max;
-                for (std::uint32_t i = local_id; i < col_count; i += local_range) {
-                    const Float cur_val = data_ptr[in_offset + i];
+                for (std::uint32_t i = local_id; i < cluster_count; i += local_range) {
```
Comment (makart19) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp:145]: types mismatch: 
i - std::uint32_t i
cluster_count - std::int64_t 
index = i (index is std::int32_t)

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp [hunk 4]
```diff
@@ -155,18 +161,24 @@ struct train_kernel_gpu<Float, method::lloyd_dense, task::clustering> {
         auto updater = cluster_updater<Float>{ queue, comm }
                            .set_cluster_count(cluster_count)
                            .set_part_count(part_count)
-                           .set_data(arr_data);
+                           .set_data(arr_data)
+                           .set_data_squares(arr_data_squares);
         updater.allocate_buffers();
 
         for (iter = 0; iter < max_iteration_count; iter++) {
```
Comment (makart19) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp:168]: why you don't declare iter here, i.e. std::int64_t iter and remove declaration on line 152, i found that you use it inside the loop only.

Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp:168]: I need it outside to report the total number of iterations

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp (hunk 2 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp]: Fixed

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp (hunk 3 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp]: Fixed

**Pull request merged**